File: backend/llm/SambaNova.go
package llm

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

var llama_405b = "Meta-Llama-3.1-405B-Instruct"

type Response struct {
	Choices []Choice `json:"choices"`
}

type Choice struct {
	Message Message `json:"message"`
}

type Message struct {
	Content string `json:"content"`
}

// free llama api
func querySambaNova(ctx context.Context, input, sys string, modelName string, temperature float32) (res string, err error) {
	url := "https://api.sambanova.ai/v1/chat/completions"
	var messages []map[string]string
	if len(sys) > 0 {
		messages = append(messages, map[string]string{"role": "system", "content": sys})
	}
	messages = append(messages, map[string]string{"role": "user", "content": input})
	requestBody := map[string]interface{}{
		"temperature": temperature,
		"messages":    messages,
		"model":       llama_405b,
	}
	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+SambaNovaApiKey)
	snClient := &http.Client{}
	resp, err := snClient.Do(req)
	if err != nil {
		return
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return
	}
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("Unexpected response status: %s\n", resp.Status)
	}
	var response Response
	err = json.Unmarshal(body, &response)
	if err != nil {
		return
	}
	if len(response.Choices) > 0 {
		return response.Choices[0].Message.Content, nil
	} else {
		return "", fmt.Errorf("No choices found in response.")
	}
}

File: backend/llm/sambanova.py
import logging
import requests

from backend.util.file import get_config

LLAMA_405B = "Meta-Llama-3.1-405B-Instruct"

def query_samba_nova(input_text: str, sys: str, model_name: str, temperature: float) -> str:
    model = get_config()['model']
    try:
        url = "https://api.sambanova.ai/v1/chat/completions"
        messages = []
        if sys:
            messages.append({"role": "system", "content": sys})
        messages.append({"role": "user", "content": input_text})

        request_body = {
            "temperature": temperature,
            "messages": messages,
            "model": model,  # Assuming model_name is passed correctly
        }
        key = get_config()['apikey']
        headers = {
            "Content-Type": "application/json",
            "Authorization": f"Bearer {key}",
        }
        response = requests.post(url, headers=headers, json=request_body)
        response.raise_for_status()  # Raises an HTTPError for bad responses

        response_data = response.json()
        if 'choices' in response_data and len(response_data['choices']) > 0:
            return response_data['choices'][0]['message']['content']
        else:
            error_message = f"No choices found in response: {response_data}"
            logging.error(error_message)
            raise ValueError(error_message)
    except requests.exceptions.HTTPError as http_err:
        error_message = f"HTTP error occurred: {http_err} - Response: {http_err.response.text}"
        logging.error(error_message)
        raise
    except requests.exceptions.ConnectionError as conn_err:
        error_message = f"Connection error occurred: {conn_err}"
        logging.error(error_message)
        raise
    except requests.exceptions.Timeout as timeout_err:
        error_message = f"Timeout error occurred: {timeout_err}"
        logging.error(error_message)
        raise
    except requests.exceptions.RequestException as req_err:
        error_message = f"Request error occurred: {req_err}"
        logging.error(error_message)
        raise
